src/main/groovy/com/twcable/grabbit/server/batch/steps/jcrnodes/JcrNodesProcessor.groovy

```diff
@@ -98,108 +97,42 @@ class JcrNodesProcessor implements ItemProcessor<JcrNode, Node> {
      * @param Mandatory childNodes
      * @return Node Proto node
      */
-    private Node buildNode(JcrNode jcrNode, Collection<JcrNode> childNodes) {
+    private Node buildNode(JCRNodeDecorator jcrNode) {
         final Node.Builder nodeBuilder = Node.newBuilder()
         nodeBuilder.setName(jcrNode.path)
         nodeBuilder.setProperties(buildProperties(jcrNode))
-        childNodes?.each {
-            nodeBuilder.addMandatoryChildNode(buildNode(it, getRequiredChildNodes(it)))
+        jcrNode.getRequiredChildNodes()?.each {
+            nodeBuilder.addMandatoryChildNode(buildNode(it))
         }
         return nodeBuilder.build()
     }
 
     /**
-     * Identify all required child nodes
-     * @param JCR Node
-     * @return list of required child nodes
-     */
-    private static Collection<JcrNode> getRequiredChildNodes(JcrNode jcrNode) {
-        return hasMandatoryChildNodes(jcrNode) ? jcrNode.getNodes().findAll{JcrNode childJcrNode -> isRequiredNode(childJcrNode)} : null
-    }
-
-    /**
-     * Checks for mandatory child nodes
-     * @param JCR Node
-     * @return boolean flag for mandatory child node
-     */
-    private static boolean hasMandatoryChildNodes(JcrNode jcrNode) {
-        return jcrNode.primaryNodeType.childNodeDefinitions.any { ((ItemDefinition)it).mandatory }
-    }
-
-    /**
-     * Checks required node
-     * @param JCR node
-     * @return boolean flag for mandatory node
-     */
-    private static boolean isRequiredNode(JcrNode node) {
-        return node.definition.isMandatory()
-    }
-
-    /**
      * Build Proto Property object from JCR node
      * @param jcrNode for Proto properties
      * @return Proto Property object
      */
-    private static NodeProtos.Properties buildProperties(JcrNode jcrNode) {
-        final List<JcrProperty> properties = jcrNode.properties.toList()
+    private static NodeProtos.Properties buildProperties(JCRNodeDecorator jcrNode) {
```

Comment:
I think we can actually eliminate this static method entirely with your addition of JCRPropertyDecorator. 

Every node, already knows about it's own properties; so each JCRNodeDecorator, should be able to wrap and build it's own JCRPropertyDecorator objects. In addition, if you go forward with the comment above this one; each property should be able to inspect itself and know if it should be serialized/transferred.